python/lsst/analysis/drp/gatherResourceUsage.py

```diff
@@ -0,0 +1,550 @@
+              configured resource usage statistics.
+        """
+        dimensions = universe.extract(self.config.dimensions)
+        # Transform input list into a dict keyed by data ID.
+        handles_by_data_id = {}
+        for handle in input_metadata:
+            handles_by_data_id[handle.dataId] = handle
+        n_rows = len(handles_by_data_id)
+        # Create a dict of empty column arrays that we'll ultimately make into
+        # a table.
+        columns = {
+            d.name: np.zeros(n_rows, dtype=_dtype_from_field_spec(d.primaryKey))
+            for d in dimensions
+        }
+        for attr_name in ("memory", "prep_time", "init_time", "run_time"):
+            if getattr(self.config, attr_name):
+                columns[attr_name] = np.zeros(n_rows, dtype=float)
+        for method_name in self.config.method_times:
+            columns[method_name] = np.zeros(n_rows, dtype=float)
+        # Populate the table, one row at a time.
+        warned_about_metadata_version = False
+        for index, (data_id, handle) in enumerate(handles_by_data_id.items()):
+            # Fill in the data ID columns.
+            for k, v in data_id.full.byName().items():
+                columns[k][index] = v
+            # Load the metadata dataset and fill in the columns derived from
+            # it.
+            metadata = handle.get()
+            try:
+                quantum_metadata = metadata["quantum"]
+            except KeyError:
+                self.log.warning(
+                    "Metadata dataset %s @ %s has no 'quantum' key.",
+                    handle.ref.datasetType.name,
+                    handle.dataId,
+                )
+            else:
+                if self.config.memory:
+                    columns["memory"][index], warned_about_metadata_version = self._extract_memory(
+                        quantum_metadata,
+                        handle,
+                        warned_about_metadata_version,
+                    )
+                for key, value in self._extract_quantum_timing(quantum_metadata).items():
+                    columns[key][index] = value
+            for key, value in self._extract_method_timing(metadata, handle).items():
+                columns[key][index] = value
+        return Struct(output_table=pd.DataFrame(columns, copy=False))
+
+    def _extract_memory(self, quantum_metadata, handle, warned_about_metadata_version):
+        """Extract maximum memory usage from quantum metadata.
+
+        Parameters
+        ----------
+        quantum_metadata : `lsst.pipe.base.TaskMetadata`
+            The nested metadata associated with the label "quantum" inside a
+            PipelineTask's metadata.
+        handle : `lsst.daf.butler.DeferredDatasetHandle`
+            Butler handle for the metadata dataset; used to identify the
+            metadata in diagnostic messages only.
+        warned_about_metadata_version : `bool`
```

Comment:
For some reason I thought warnings themselves could be configured to only be emitted once

Reply:
If so I'm not familiar with it, and the message does include the data ID so the messages wouldn't be identical anyway.